components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt

```
@@ -462,6 +464,9 @@ class DynamicMemberRegistrationService @Activate constructor(
                     require(isNotEmpty()) { "No notary key ID was provided." }
                     require(orderVerifier.isOrdered(this, 3)) { "Provided notary key IDs are incorrectly numbered." }
                 }
+                context.keys.filter { notaryProtocolVersionsRegex.matches(it) }.apply {
+                    require(orderVerifier.isOrdered(this, 6)) { "Provided notary protocol versions are incorrectly numbered." }
```

can this be an empty list?

---

I believe it would be empty if the protocol is not specified

---

but must you specify the protocol when you create a notary?

---

also I can see that in the converter we throw an exception if it was empty:
`val protocolVersions = context.map.parseList(PROTOCOL_VERSIONS_PREFIX, Int::class.java).toSet().apply {
             require(isNotEmpty()) { throw ValueNotFoundException("'$PROTOCOL_VERSIONS_PREFIX' is empty.") }
         }`
        wouldn't it be better to throw this exception here? or if it can be empty then default to emptyList() in the notary info converter
        (although I think parseList automatically returns empty list if the key wasn't found)

---

My understanding is that it is tied to the protocol name (which is not mandatory when adding a notary to an existing notary service). So if protocol name is not provided, versions can be missing in the registration context. However while adding a new notary service, protocol name is required (hence, protocol version is required as well) - so versions shouldn't be an empty list in `NotaryInfo`